sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -3968,3 +3964,267 @@ object ArrayUnion {
+              s"get($i, $et)", s"update($pos, $value)", "Object",
+              s"${ev.value} = new $genericArrayData(new Object[$size]);")
+        }
+      } else {
+        ("", "", "", "", "", "", "", "")
+      }
+
+    nullSafeCodeGen(ctx, ev, (array1, array2) => {
+      if (openHashElementType != "") {
+        // Here, we ensure elementTypeSupportEquals is true
+        val notFoundNullElement = ctx.freshName("notFoundNullElement")
+        val openHashSet = classOf[OpenHashSet[_]].getName
+        val classTag = s"scala.reflect.ClassTag$$.MODULE$$.$openHashElementType()"
+        val hs = ctx.freshName("hs")
+        val arrayData = classOf[ArrayData].getName
+        val arrays = ctx.freshName("arrays")
+        val array = ctx.freshName("array")
+        val arrayDataIdx = ctx.freshName("arrayDataIdx")
+
+        val array2NullCheck = if (right.dataType.asInstanceOf[ArrayType].containsNull) {
+          s"""
+            |if ($array2.isNullAt($i)) {
+            |  $notFoundNullElement = false;
+            |} else
+           """.stripMargin
+        } else {
+          ""
+        }
+        val array1NullCheck = if (left.dataType.asInstanceOf[ArrayType].containsNull) {
+          s"""
+             |if ($array1.isNullAt($i)) {
+             |  if ($notFoundNullElement) {
+             |    $size++;
+             |    $notFoundNullElement = false;
+             |  }
+             |} else
+           """.stripMargin
+        } else {
+          ""
+        }
+        val array1NullAssignment = if (left.dataType.asInstanceOf[ArrayType].containsNull) {
+          s"""
+             |if ($array1.isNullAt($i)) {
+             |  if ($notFoundNullElement) {
+             |    ${ev.value}.setNullAt($pos++);
+             |    $notFoundNullElement = false;
+             |  }
+             |} else
+           """.stripMargin
+        } else {
+          ""
+        }
+
+        s"""
+           |$openHashSet $hs = new $openHashSet$postFix($classTag);
+           |boolean $notFoundNullElement = true;
+           |int $size = 0;
+           |for (int $i = 0; $i < $array2.numElements(); $i++) {
+           |  $array2NullCheck
+           |  {
+           |    $javaTypeName $value = $array2.$getter;
+           |    $hsJavaTypeName $hsValue = $genHsValue;
+           |    $hs.add$postFix($hsValue);
+           |  }
+           |}
+           |for (int $i = 0; $i < $array1.numElements(); $i++) {
+           |  $array1NullCheck
+           |  {
+           |    $javaTypeName $value = $array1.$getter;
+           |    $hsJavaTypeName $hsValue = $genHsValue;
+           |    if (!$hs.contains($hsValue)) {
+           |      $hs.add$postFix($hsValue);
+           |      $size++;
+           |    }
+           |  }
+           |}
+           |$arrayBuilder
+           |$hs = new $openHashSet$postFix($classTag);
+           |$notFoundNullElement = true;
+           |int $pos = 0;
+           |for (int $i = 0; $i < $array2.numElements(); $i++) {
```

Comment:
why add `array2` to the hash set again?

Reply:
We add some elements in `array1`to the hash set at L4192.
Since we want to create a hashset that have elements only of `array2`, we reallocate a hashset at L4198 and add elements of `array2` again here.

Reply:
then why we do the first iteration? only calculating the size?

Reply:
Yes, only calculating array size for allocating a `UnsafeArrayData` or `GenericArrayData`.

Reply:
is this worth? I feel using an `ArrayBuffer` like the interpreted path is faster.

Reply:
To use 'ArrayBuffer' involves boxing. Let me consider 'ArrayBuffer.ofInt' or others with special null handling.